### ISSUE TITLE
Patch Entitlements through User endpoint

### ISSUE THREAD

**Issue by Biggo89** [simpleidserver/SimpleIdServer#153]
Hello,
I have implemented a new Entitlements Resource exactly how has been done for Groups Resource, so to add an entitlements to a specific user we call Entitlements endpoint patching members attribute.

In my application I have a third part that want patch entitlements through Users endpoint, see below schema for expected request and response:

![image](https://user-images.githubusercontent.com/11877326/121550514-16c10200-ca0f-11eb-8370-e16f534573f8.png)
![image](https://user-images.githubusercontent.com/11877326/121556354-18d98f80-ca14-11eb-8289-9e835aaef66e.png)

I tried to change some Configurations on SCIMAttributeMapping and UserSchema like below:

UserSchema

![image](https://user-images.githubusercontent.com/11877326/121551343-d6ae4f00-ca0f-11eb-8de5-9ed9ebe6a647.png)

SCIMAttributeMapping

![image](https://user-images.githubusercontent.com/11877326/121551477-f9406800-ca0f-11eb-8c27-a0b3fbafa3d4.png)

But it doesn't work as expected, below the result after some tests:

Patch User

![image](https://user-images.githubusercontent.com/11877326/121554533-85538f00-ca12-11eb-9e04-38c9d46bacb4.png)

Get User

![image](https://user-images.githubusercontent.com/11877326/121554743-b2a03d00-ca12-11eb-9103-4606d8b14e8f.png)

Get Entitlements

![image](https://user-images.githubusercontent.com/11877326/121554855-cb105780-ca12-11eb-81ab-da856c4b4145.png)

Can you give me the reason why it is not working as expected since the behavior is not matching the schema I linked on top?

**Comment by simpleidserver** [comment]
Hello,

Your configuration is almost correct :). 
If you want to assign "entitlements" to a user, you must execute a PATCH operation against the "Entitlements" resource and not against the "User" resource. 
The behavior is similar to "groups" assignment. Groups are assigned to users by executing HTTP PATCH operations against the "Groups" endpoint (for more information you can refer to the RFC : https://datatracker.ietf.org/doc/html/rfc7644#section-3.5.2.1).

**Modifications made on the C# project** :

I made the following modifications to have a working example :

1. Create entitlement schema :

            var entitlementSchema = Builder.SCIMSchemaBuilder.Create("urn:entitlement", "Entitlement", "Entitlements", string.Empty, true)
                .AddStringAttribute("displayName")
                .AddComplexAttribute("members", opt =>
                {
                    opt.AddStringAttribute("value");
                    opt.AddStringAttribute("$ref");
                    opt.AddStringAttribute("type");
                }, multiValued: true)
                .Build();
		
2. Create custom user resource :

            var customUserSchema = Builder.SCIMSchemaBuilder.Create("urn:customuser", "CustomUser", "CustomUsers", string.Empty, true)
                .AddStringAttribute("userName", required: true)
                .AddComplexAttribute("entitlements", opt =>
                {
                    opt.AddStringAttribute("value", mutability: SCIMSchemaAttributeMutabilities.READONLY);
                    opt.AddStringAttribute("$ref", mutability: SCIMSchemaAttributeMutabilities.READONLY);
                    opt.AddStringAttribute("type", mutability: SCIMSchemaAttributeMutabilities.READONLY);
                }, multiValued: true)
                .Build();
				
3. Add two new controllers : CustomUserController & CustomResourcesController
4. Add a new Attribute mapping :

                new SCIMAttributeMapping
                {
                    Id = Guid.NewGuid().ToString(),
                    SourceAttributeId = customUserSchema.Attributes.First(a => a.Name == "entitlements").Id,
                    SourceResourceType = "CustomUsers",
                    SourceAttributeSelector = "entitlements",
                    TargetResourceType = "Entitlements",
                    TargetAttributeId = entitlementSchema.Attributes.First(a => a.Name == "members").SubAttributes.First(a => a.Name == "value").Id
                }
				
The project **SimpleIdServer.Scim.Startup** (branch release/1.2.2) has been updated with those changes (please check [here](https://github.com/simpleidserver/SimpleIdServer/tree/release/1.2.2/src/Scim/SimpleIdServer.Scim.Startup)).

You can run the **SimpleIdServer.Scim.Startup** project and execute the following HTTP REQUESTS :

1. Create custom user

```
HTTP POST : http://localhost:60002/CustomUsers
{
    "schemas": ["urn:customuser"],
    "userName": "userName14a4979b3984"
}
```

2. Add entitlement

```
HTTP POST : http://localhost:60002/Entitlements
{
    "schemas": ["urn:entitlement"],
    "displayName": "firstEntitlement"
}
```

3. Assign entitlement

```
HTTP PATCH : http://localhost:60002/Entitlements/{{entitlementId}}
{
    "schemas": [
        "urn:ietf:params:scim:api:messages:2.0:PatchOp"
    ],
    "Operations": [
        {
            "op": "add",
            "path": "members",
            "value": [
                {
                    "value": "{{customUserId}}"
                }
            ]
        }
    ]
}
```

4. Get user :

`HTTP GET : http://localhost:60002/CustomUsers/{{customUserId}}`

**Comment by simpleidserver** [comment]
There is probably an issue in the "AttributeReferenceEnricher", i'm going to fix it.
Sorry for the inconvenience

**Comment by simpleidserver** [comment]
The issue should be fixed in the branch "release/1.2.2", can-you please run the project [SimpleIdServer.Scim.Startup](https://github.com/simpleidserver/SimpleIdServer/tree/release/1.2.2/src/Scim/SimpleIdServer.Scim.Startup) & check if it's working as expected ?  Entitlements can be added via the Users endpoint.

**Comment by gabrielemilan** [comment]
Hello, we tried the fix and we found another issue. If the user has more than one entitlements only the first one has $ref and display property filled. We found this piece of code, that takes the first element of entitlements:

![image](https://user-images.githubusercontent.com/10979448/122232320-6f831580-cebb-11eb-9de0-67c497a04e27.png)

the result is:

![image](https://user-images.githubusercontent.com/10979448/122232586-a1947780-cebb-11eb-8139-e244efca92d9.png)

Can you check please?

**Comment by simpleidserver** [comment]
Hello,

The bug is fixed in the branch "release/1.2.2"
